Modal Images

sbddesign opened the pull request:
I added the ability to define a modal view for images, which is outlined in the Formatting page. This is a proposed fix for #267.

🖼 [Deploy Preview](https://deploy-preview-629--sad-borg-390916.netlify.app/guide/contribute/formatting/#image-viewable-in-larger-modal)

johnsBeharry commented on guide/contribute/formatting.md:
```diff
@@ -175,6 +175,36 @@ Let's start with a very wide image that extends beyond the content width on desk
+   mobile = "/assets/images/guide/contribute/formatting/example-optional-image-modal-mobile.png"
+   mobileRetina = "/assets/images/guide/contribute/formatting/example-optional-image-modal-mobile@2x.png"
+   modalImage = "/assets/images/guide/contribute/formatting/example-optional-image-modal@2x.png"
+   modalCaption = "An optional caption to display with the image"
```
The picture include already has a caption, curious what's the rationale of adding a specific parameter for modal caption. 👇  is a "caption" on the picture.html include

<img width="217" alt="Screenshot 2022-01-22 at 1 30 12 PM" src="https://user-images.githubusercontent.com/183140/150649200-2cff167f-cf20-4f4f-9856-95de02113fe6.png">

sbddesign replied:
There was some talk in #267 about being able to provide additional information with each image. But it could easily just re-use the same caption.

sbddesign replied:
@GBKS and @Bosch-0, any thoughts on this? Should there be a separate caption for modal images, or should it just use the same caption as the original image?

GBKS replied:
I'd not introduce a new caption. Reason is that this would lead to content hidden from a "normal" page read-through.

Overall, I think this image zoom should not expose new information, but should make the information on the page (in the image) just easier to see.

Bosch-0 replied:
Agree with GBKS, a new caption seems unnecessary.

GBKS reviewed (changes requested):
Thanks for tackling this one.

It would be ideal if images that can be viewed in modals had some type of indication. Otherwise users have to guess if they can zoom in on an image.

In the modal, when the image is loading, the close button and the text are centered in the screen, which looks broken. Once the image is loaded, things pop in place. Would be best to either provide width/height ahead of time so the layout calculation can be done before the image is loaded, or hide things (show a loader?) until the image is loaded.

Another issue is that vertical resizing doesn't work well. The image will always stretch to fill the screen width (minus padding). That is fine when the image height is smaller than the screen height. Otherwise, the image is cut off at the top and bottom because it extends beyond screen edges. You may also need image width/height properties to handle this properly. I see you're using the width/height from the img tag, but that will only be accurate for desktop.

I'd also make the "x" in the close button smaller and white and add a slight hover animation.

What do you think of adding the option of adding a mobile and a desktop modal image? I think diagrams are a prime use case for this functionality, and oftentimes we have two versions of diagrams for appropriate viewing on landscape and portrait screens.

Hope that's helpful and all makes sense.

sbddesign commented:
Thanks for the feedback, @GBKS. Yeah, I can take care of all this.

> Another issue is that vertical resizing doesn't work well.

There is a little bit of built-in logic for handling different aspect ratios that kicks in when the page first loads. I can probably just have it recalculate this on a window resize.

> What do you think of adding the option of adding a mobile and a desktop modal image?

Yeah, seems reasonable.

sbddesign commented:
@GBKS I pushed several updates.
- The modal uses the same caption as the main image (@johnsBeharry)
- Added better load handling with some animation (try using throttling in dev inspector to test)
- Added better screen resize handling
- Added ability to define alternate modal image for mobile (the [landscape image on the Formatting page](https://deploy-preview-629--sad-borg-390916.netlify.app/guide/contribute/formatting/#image-viewable-in-larger-modal-with-alternate-mobile-image) is an example of this).

For adding an indicator that an image has a modal, I'm trying to think what the best way to do this will be. The two approaches I have so far are 1) adding a little "expand" icon next to the caption and 2) adding the expand icon as an overlay in the corner of the image. Option 2 runs the risk of cluttering the image with extra UI, and the icon not laways looking good on every image. So I favor option 1. In situations where no caption is defined, I can simply inject a default caption like "expand image". [See an example here.](https://www.figma.com/file/Kg6vEZHRUo3AFFGdqjA9Qe/Image-Modal-Indicators?node-id=0%3A1)

Thoughts?

sbddesign commented:
Also, I went against your suggestion a little @GBKS on the "X". Initially, it did look good when I made it small like you asked. Long-story-short, to accommodate better resize handling, it seemed best to detach the X from the image and let it remain persistently in the upper-right hand corner. Once the "X" becomes detached, it needs to be a little larger to be visible. By I took the circle background away so that it would be a simpler line icon like the other icons on the Bitcoin Design site. How do you feel about this?

Bosch-0 reviewed (approved):
This LGTM, very clean - love the loading animation.

GBKS commented:
Looks super good, thank you for addressing everything. I like the idea of adding the "Expand image" caption automatically. It's simple and does not interfere with images.

Only addition I can think of is disabling page scroll when the modal is visible. Not a biggy though.

pavlenex reviewed (approved):
LGTM, great idea, just did a small commit to resolve an img conflict.

sbddesign commented:
Thanks for the merge -- I was actually still working on some feedback from Christoph though, I was trying to get his OK on the Figma design before I merged it into the code. I'll open up a new PR today.
